onnxruntime/core/providers/webnn/builders/model.cc

```
@@ -103,15 +127,34 @@ void Model::AllocateInputOutputBuffers() {
     const auto& input_info = input_output_info_.at(input);
     const auto input_shape = input_info.shape;
     const auto num_elements = SafeInt<size_t>(Product(input_shape));
-    wnn_inputs_.set(input,
-                    emscripten::val::global("Float32Array").new_(static_cast<const int>(num_elements)));
+    const auto data_type = input_info.data_type;
+    switch (data_type) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT16:
+        wnn_inputs_.set(input,
+                        emscripten::val::global("Uint16Array").new_(static_cast<const uint16_t>(num_elements)));
+        break;
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+        wnn_inputs_.set(input,
+                        emscripten::val::global("Float32Array").new_(static_cast<const float>(num_elements)));
+        break;
+      default:
+        break;
```

I'd bail here on an unsupported data type rather than silently continuing, probably calling `ORT_THROW`.